### PR TITLE
Replace yarn with npm

### PR DESCRIPTION
**Pull request by robin-da**
We replace `yarn` with `npm` and update the daml assistant and compatibility tests accordingly. To update the @daml.js library, just rerun the codegen, npm will pick up the changes automatically.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by nemanja-da — APPROVED** [review]
Thnx Robin!

File: compatibility/bazel_tools/create-daml-app/Main.hs [hunk 1]
```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
...
-            [(DamlLedger, damlLedgerPath), (DamlReact, damlReactPath), (DamlTypes, damlTypesPath)]
+          setupNpmEnv uiDir [(DamlTypes, damlTypesPath)
+                            , (DamlLedger, damlLedgerPath)
+                            , (DamlReact, damlReactPath)
+                            ]
           step "Install Jest, Puppeteer and other dependencies"
           addTestDependencies "package.json" testDepsPath
-          retry 3 (callProcessSilent yarnPath ["install"])
+          patchTsDependencies uiDir "package.json"
+          -- use '--scripts-prepend-node-path' to make sure we are using the correct 'node' binary
```
Comment (cocreature) [inline compatibility/bazel_tools/create-daml-app/Main.hs:216]: Which one does it use if you don’t do that?

File: compatibility/bazel_tools/create-daml-app/Main.hs [hunk 2]
```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
...
+          retry 3 (callProcessSilent npmPath ["install", "--scripts-prepend-node-path"])
           step "Run Puppeteer end-to-end tests"
-          copyFile testTsPath (cdaDir </> "ui" </> "src" </> "index.test.ts")
-          callProcess yarnPath ["run", "test", "--ci", "--all"]
+          copyFile testTsPath (uiDir </> "src" </> "index.test.ts")
+          -- we need 'npm-cli.js' in the path for the following test
```
Comment (cocreature) [inline compatibility/bazel_tools/create-daml-app/Main.hs:220]: Why do we need this in this test but I don’t need seem to have it in PATH on my machine where I’ve installed NPM?

File: compatibility/bazel_tools/create-daml-app/Main.hs [hunk 3]
```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
...
+    let uiLibPath = uiDir </> name
+    extractTarGz path uiLibPath
+    patchTsDependencies uiDir $ uiLibPath </> "package.json"
+
+-- | Overwrite dependencies to our TypeScript libraries to point to local file dependencies in the
+-- 'ui' directory in the specified package.json file.
+patchTsDependencies :: FilePath -> FilePath -> IO ()
+patchTsDependencies uiDir packageJsonFile = do
+  packageJson0 <- readJsonFile packageJsonFile
+  case packageJson0 of
```
Comment (cocreature) [inline compatibility/bazel_tools/create-daml-app/Main.hs:260]: I like `lens-aeson` for this kind of ad-hoc JSON manipulation but I’ll leave it up to you if you want to use it.

File: compatibility/bazel_tools/create-daml-app/Main.hs [hunk 4]
```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
...
+          -- patch dependencies to point to local files if they are present in the package.json
+          let patchedDeps =
+                HMS.fromList
+                  ([ ( "@daml.js/create-daml-app"
+                     , Aeson.String $
+                       T.pack $
+                       "file:" <> "./daml.js/create-daml-app-0.1.0")
+                   | "@daml.js/create-daml-app" `elem` depNames
+                   ] ++
+                   [ (depName, Aeson.String $ T.pack $ "file:" <> libRelPath)
```
Comment (cocreature) [inline compatibility/bazel_tools/create-daml-app/Main.hs:274]: technically relative file urls are invalid. I guess npm handles them given that this test passes but perhaps worth keeping in mind.

File: compatibility/bazel_tools/create-daml-app/Main.hs [hunk 5]
```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
...
+                   ]) `HMS.union`
+                dependencies
+          let newPackageJson =
+                Aeson.Object $
+                HMS.insert "dependencies" (Aeson.Object patchedDeps) packageJson
+          -- Make sure we have write permissions before writing
+          p <- getPermissions packageJsonFile
+          setPermissions packageJsonFile (setOwnerWritable True p)
+          BSL.writeFile packageJsonFile (Aeson.encode newPackageJson)
+        Nothing -> pure () -- Nothing to patch
```
Comment (cocreature) [inline compatibility/bazel_tools/create-daml-app/Main.hs:291]: Do we ever expect to hit this case?

File: templates/create-daml-app/ui/package.json.template [hunk 6]
```diff
@@ -3,7 +3,7 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
-    "@daml.js/__PROJECT_NAME__": "file:../daml.js/__PROJECT_NAME__-0.1.0",
+    "@daml.js/__PROJECT_NAME__": "file:daml.js/__PROJECT_NAME__-0.1.0",
```
Comment (cocreature) [inline templates/create-daml-app/ui/package.json.template:6]: Does `../daml.js` not work with npm or why did you change this?

File: docs/source/getting-started/index.rst [hunk 7]
```diff
@@ -69,12 +62,12 @@ Once the DAR file is created you will see this message in terminal ``Created .da
...
 
-    daml codegen js .daml/dist/create-daml-app-0.1.0.dar -o daml.js
+    daml codegen js .daml/dist/create-daml-app-0.1.0.dar -o ui/daml.js
 
-Now, changing to the ``ui`` folder, use Yarn to install the project dependencies::
+Now, changing to the ``ui`` folder, use ``npm`` to install the project dependencies::
 
     cd ui
-    yarn install --force --frozen-lockfile
+    npm install --force --frozen-lockfile
```
Comment (cocreature) [inline docs/source/getting-started/index.rst:70]: I assume you tested that `--force --frozen-lockfile` works on npm?

**Review by cocreature — COMMENTED** [review]
Nice work, thank you!

This also needs to update the release instructions which currently still reference yarn.

**Comment by nemanja-da** [comment]
@cocreature think this is still all WiP, right @robin-da?

**Comment by cocreature** [comment]
Screenshot from 2020-09-07 14-10-57
It was a WIP until this morning.

File: templates/create-daml-app/ui/package.json.template (hunk 6 above)
Comment (robin-da) [reply templates/create-daml-app/ui/package.json.template:6]: No, we need the generated code in the root level of the ui package.

File: docs/source/getting-started/index.rst (hunk 7 above)
Comment (robin-da) [reply docs/source/getting-started/index.rst:70]: yep

File: compatibility/bazel_tools/create-daml-app/Main.hs (hunk 5 above)
Comment (robin-da) [reply compatibility/bazel_tools/create-daml-app/Main.hs:291]: That would be the rare case when there are actually no dependencies. Not sure what kind of `package.json` files we generate.

File: compatibility/bazel_tools/create-daml-app/Main.hs (hunk 4 above)
Comment (robin-da) [reply compatibility/bazel_tools/create-daml-app/Main.hs:274]: I didn't know that! We did have absolute file path in there first and changed it to relative ones.

File: compatibility/bazel_tools/create-daml-app/Main.hs (hunk 1 above)
Comment (robin-da) [reply compatibility/bazel_tools/create-daml-app/Main.hs:216]: can't quite remember whether it tried a different path or couldn't find it at all.

File: compatibility/bazel_tools/create-daml-app/Main.hs (hunk 2 above)
Comment (robin-da) [reply compatibility/bazel_tools/create-daml-app/Main.hs:220]: you won't have 'npm-cli.js' in your PATH because it's a wrapper from bazel for `npm`.

**Comment by robin-da** [comment]
@nemanja-da No, now it's ready for review.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/CreateDamlAppTests.hs [hunk 8]
```diff
@@ -98,45 +105,85 @@ tests =
...
+          callCommandSilent "npm-cli.js install --force --frozen-lockfile"
           step "Run linter again"
-          callCommandSilent "yarn lint --max-warnings 0"
+          callCommandSilent "npm-cli.js run-script lint -- --max-warnings 0"
           step "Build the new UI"
-          callCommandSilent "yarn build"
+          callCommandSilent "npm-cli.js run-script build"
 
-        -- Run end to end testing for the app.
+        -- -- Run end to end testing for the app.
```
Comment (hurryabit) [inline daml-assistant/integration-tests/src/DA/Daml/Assistant/CreateDamlAppTests.hs:121]:
```suggestion
        -- Run end to end testing for the app.
```

File: templates/create-daml-app/ui/.gitignore [hunk 9]
```diff
@@ -0,0 +1 @@
+daml.js/
```
Comment (hurryabit) [inline templates/create-daml-app/ui/.gitignore:1]: I would suggest we drop this file and change the `daml.js/` entry in `templates/create-daml-app/.gitignore` into `ui/daml.js/`. That way, we need to ship less files and the project appears a tiny bit less complex.

**Review by hurryabit — APPROVED** [review]
This is awesome. Thanks a lot.

File: docs/source/getting-started/index.rst [hunk 10]
```diff
@@ -32,14 +32,7 @@ Please make sure that you have the DAML SDK, Java 8 or higher, and Visual Studio
 You will also need some common software tools to build and interact with the template project.
 
 - `Git <https://git-scm.com/downloads>`_ version control system
-- `Yarn <https://classic.yarnpkg.com/en/docs/install/>`_ package manager for JavaScript. You have to have yarn version 1.10.0 or higher.
-
-  Note: Ubuntu 17.04 and higher come with ``cmdtest`` package installed by default. If you are getting errors when installing yarn, you may want to run ``sudo apt remove cmdtest`` first and then install yarn. More information can be found `here <https://github.com/yarnpkg/yarn/issues/2821>`_ as well as in the official `yarn installation docs for Debian / Ubuntu <https://classic.yarnpkg.com/en/docs/install/#debian-stable>`_
-- `NodeJS <https://nodejs.org/en/download/>`_ in version 8.16 or higher. This will usually be
-  installed automatically as part of installing Yarn.
-
-  Note: On Ubuntu 18.04, NodeJS 8.10 will be installed as part of installing Yarn which is too old.
```
Comment (cocreature) [inline docs/source/getting-started/index.rst:41]: This comment still applies I believe. Ubuntu 18.04 ships with an ancient version of node that is incompatible with `fsevents`.

**Review by cocreature — APPROVED** [review]
Nice work thank you! One comment on the node version which still matters and I agree with @hurryabit wtr to the `.gitignore`.

File: templates/create-daml-app/ui/.gitignore (hunk 9 above)
Comment (robin-da) [reply templates/create-daml-app/ui/.gitignore]: changed it, thanks!

File: docs/source/getting-started/index.rst (hunk 10 above)
Comment (robin-da) [reply docs/source/getting-started/index.rst:41]: ok, I'll put it back in without the yarn part.

File: compatibility/bazel_tools/create-daml-app/Main.hs (hunk 4 above)
Comment (garyverhaegen-da) [reply compatibility/bazel_tools/create-daml-app/Main.hs:274]: "Technically" as in the RFC for URLs? NPM seems to be fine with them [per their documentation of package.json](https://docs.npmjs.com/files/package.json#local-paths).

Comment (cocreature) [reply compatibility/bazel_tools/create-daml-app/Main.hs:274]: yep, according to the rfc you need a leading slash.
